File: scripts/serve-site.mjs
import { createServer } from "node:http";
import { createReadStream } from "node:fs";
import { stat } from "node:fs/promises";
import { dirname, extname, relative, resolve, sep } from "node:path";
import { networkInterfaces } from "node:os";
import { fileURLToPath } from "node:url";

const root = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const args = parseArgs(process.argv.slice(2));
const siteRoot = resolve(root, args.root ?? ".");
const host = args.host ?? "0.0.0.0";
const requestedPort = Number(args.port ?? 8000);

const mimeTypes = new Map([
  [".html", "text/html; charset=utf-8"],
  [".css", "text/css; charset=utf-8"],
  [".js", "text/javascript; charset=utf-8"],
  [".mjs", "text/javascript; charset=utf-8"],
  [".json", "application/json; charset=utf-8"],
  [".svg", "image/svg+xml"],
  [".png", "image/png"],
  [".jpg", "image/jpeg"],
  [".jpeg", "image/jpeg"],
  [".gif", "image/gif"],
  [".webp", "image/webp"],
  [".ico", "image/x-icon"],
  [".txt", "text/plain; charset=utf-8"],
  [".map", "application/json; charset=utf-8"],
]);

start(requestedPort);

function start(port) {
  const server = createServer(handleRequest);

  server.on("error", (error) => {
    if (error.code === "EADDRINUSE" || error.code === "EACCES") {
      start(port + 1);
      return;
    }

    throw error;
  });

  server.listen(port, host, () => {
    const activePort = server.address().port;

    console.log(`Serving ${siteRoot}`);
    console.log(`Local:   http://127.0.0.1:${activePort}/`);
    for (const address of getLanAddresses()) {
      console.log(`Network: http://${address}:${activePort}/`);
    }
  });
}

async function handleRequest(request, response) {
  if (request.method !== "GET" && request.method !== "HEAD") {
    response.writeHead(405, { Allow: "GET, HEAD" });
    response.end("Method Not Allowed");
    return;
  }

  try {
    const filePath = await resolveRequestPath(request.url);
    const fileStat = await stat(filePath);

    if (!fileStat.isFile()) {
      response.writeHead(404, { "Content-Type": "text/plain; charset=utf-8" });
      response.end("Not Found");
      return;
    }

    response.writeHead(200, {
      "Content-Type": mimeTypes.get(extname(filePath).toLowerCase()) ?? "application/octet-stream",
      "Content-Length": fileStat.size,
      "Cache-Control": "no-store",
    });

    if (request.method === "HEAD") {
      response.end();
      return;
    }

    createReadStream(filePath).pipe(response);
  } catch (error) {
    if (error.code === "ENOENT") {
      response.writeHead(404, { "Content-Type": "text/plain; charset=utf-8" });
      response.end("Not Found");
      return;
    }

    response.writeHead(500, { "Content-Type": "text/plain; charset=utf-8" });
    response.end("Internal Server Error");
    console.error(error);
  }
}

function getLanAddresses() {
  return Object.values(networkInterfaces())
    .flat()
    .filter((details) => details?.family === "IPv4" && !details.internal)
    .map((details) => details.address);
}

async function resolveRequestPath(requestUrl) {
  const url = new URL(requestUrl, "http://localhost");
  const pathname = decodeURIComponent(url.pathname);
  const normalizedPath = pathname.endsWith("/") ? `${pathname}index.html` : pathname;
  const requestedPath = resolve(siteRoot, `.${normalizedPath}`);
  const relativePath = relative(siteRoot, requestedPath);

  if (relativePath.startsWith("..") || relativePath === "" || relativePath.split(sep).includes("..")) {
    const error = new Error("Forbidden path");
    error.code = "ENOENT";
    throw error;
  }

  return requestedPath;
}

function parseArgs(rawArgs) {
  const parsed = {};

  for (let index = 0; index < rawArgs.length; index++) {
    const arg = rawArgs[index];
    if (!arg.startsWith("--")) continue;

    const key = arg.slice(2);
    const next = rawArgs[index + 1];
    if (!next || next.startsWith("--")) {
      parsed[key] = true;
      continue;
    }

    parsed[key] = next;
    index++;
  }

  return parsed;
}

File: scripts/serve-site.py
from __future__ import annotations

import argparse
import errno
import functools
import http.server
import socket
from pathlib import Path


MIME_TYPES = {
    ".html": "text/html; charset=utf-8",
    ".css": "text/css; charset=utf-8",
    ".js": "text/javascript; charset=utf-8",
    ".mjs": "text/javascript; charset=utf-8",
    ".json": "application/json; charset=utf-8",
    ".svg": "image/svg+xml",
    ".png": "image/png",
    ".jpg": "image/jpeg",
    ".jpeg": "image/jpeg",
    ".gif": "image/gif",
    ".webp": "image/webp",
    ".ico": "image/x-icon",
    ".txt": "text/plain; charset=utf-8",
    ".map": "application/json; charset=utf-8",
}


class TimeScaleRequestHandler(http.server.SimpleHTTPRequestHandler):
    def guess_type(self, path: str) -> str:
        suffix = Path(path).suffix.lower()
        return MIME_TYPES.get(suffix, super().guess_type(path))

    def end_headers(self) -> None:
        self.send_header("Cache-Control", "no-store")
        super().end_headers()

    def log_message(self, format: str, *args) -> None:
        return


class ReusableThreadingHTTPServer(http.server.ThreadingHTTPServer):
    allow_reuse_address = True


def main() -> int:
    args = parse_args()
    root = Path(args.root).resolve()

    if not root.exists():
        print(f"Site root does not exist: {root}")
        print("Run `npm run build` first, or use `serve-site.bat`.")
        return 1

    handler = functools.partial(TimeScaleRequestHandler, directory=str(root))
    server, port = bind_server(handler, args.host, args.port)

    print(f"Serving {root}", flush=True)
    print(f"Local:   http://127.0.0.1:{port}/", flush=True)
    for address in get_lan_addresses():
        print(f"Network: http://{address}:{port}/", flush=True)
    print("Use a Network URL from your phone. Ctrl+C stops the server.", flush=True)

    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nServer stopped.", flush=True)
    finally:
        server.server_close()

    return 0


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Serve To Scale: Time with browser-safe MIME types.")
    parser.add_argument("--root", default="_site", help="Directory to serve.")
    parser.add_argument("--host", default="0.0.0.0", help="Host/interface to bind.")
    parser.add_argument("--port", default=8126, type=int, help="Preferred port. The server will try nearby ports if needed.")
    return parser.parse_args()


def bind_server(handler: type[http.server.SimpleHTTPRequestHandler], host: str, start_port: int):
    last_error: OSError | None = None

    for port in range(start_port, start_port + 50):
        try:
            return ReusableThreadingHTTPServer((host, port), handler), port
        except OSError as error:
            last_error = error
            if error.errno not in {errno.EADDRINUSE, errno.EACCES, 10013, 10048}:
                raise

    raise RuntimeError(f"Could not bind {host}:{start_port}-{start_port + 49}") from last_error


def get_lan_addresses() -> list[str]:
    addresses = set()

    try:
        hostname = socket.gethostname()
        for address in socket.gethostbyname_ex(hostname)[2]:
            if not address.startswith("127."):
                addresses.add(address)
    except OSError:
        pass

    try:
        with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
            sock.connect(("8.8.8.8", 80))
            address = sock.getsockname()[0]
            if not address.startswith("127."):
                addresses.add(address)
    except OSError:
        pass

    return sorted(addresses)


if __name__ == "__main__":
    raise SystemExit(main())
